### PR TITLE
Add integration test log scanning and fix PRT duplicate join race condition

### PR DESCRIPTION
**Pull request by vfusco**

### CODE REVIEW

File: test/integration/cli_helpers_test.go [hunk 1]
```diff
@@ -132,6 +132,39 @@ func deployApplication(ctx context.Context, appName, dappPath string, extraArgs
...
+	out, err := runCLI(ctx, "inspect", appName, payload)
+	if err != nil {
+		return nil, fmt.Errorf("inspect %s: %w", appName, err)
+	}
+	var result inspectResult
+	if err := json.Unmarshal([]byte(out), &result); err != nil {
+		return nil, fmt.Errorf("parse inspect output: %w", err)
+	}
+	return &result, nil
+}
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/cli_helpers_test.go:166]: inspect is only json.

File: test/compose/compose.integration.yaml [hunk 2]
```diff
@@ -77,7 +77,7 @@ services:
   node:
     image: cartesi/rollups-node:devel
     init: true
-    command: cartesi-rollups-node
+    command: ["sh", "-c", "cartesi-rollups-node 2>&1 | tee /logs/node.log"]
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/compose/compose.integration.yaml]: fixed

File: test/integration/cli_helpers_test.go [hunk 3]
```diff
@@ -132,6 +132,39 @@ func deployApplication(ctx context.Context, appName, dappPath string, extraArgs
 	return app.IApplicationAddress, nil
 }
 
+// disableApplication sets the application status to disabled so the node
+// stops processing it. Call this in TearDownSuite to prevent a finished
+// test's application from generating errors during subsequent suites.
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/cli_helpers_test.go]: fixed

File: test/integration/logscanner_test.go [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+		}
+		if ts.Before(windowStart) || ts.After(windowEnd) {
+			continue
+		}
+		if idx := matchedAllowedIdx(line, allowed); idx >= 0 {
+			matched[idx] = true
+		} else {
+			unexpected = append(unexpected, line)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/logscanner_test.go:154]: fixed

File: test/integration/echo_authority_test.go [hunk 5]
```diff
@@ -31,16 +34,83 @@ func (s *EchoAuthoritySuite) TearDownSuite() {
...
+		result, err = inspectApplication(s.ctx, s.appName, "hello")
+		if err == nil {
+			break
+		}
+		select {
+		case <-deadline:
+			s.Require().NoError(err, "inspect did not become available within timeout")
+		case <-time.After(2 * time.Second):
+		}
+	}
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/echo_authority_test.go:96]: fixed

File: test/compose/compose.integration.yaml [hunk 6]
```diff
@@ -77,7 +77,7 @@ services:
   node:
     image: cartesi/rollups-node:devel
     init: true
-    command: cartesi-rollups-node
+    command: ["sh", "-c", "set -o pipefail; cartesi-rollups-node 2>&1 | tee /logs/node.log"]
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/compose/compose.integration.yaml]: fixed

File: test/integration/logscanner_test.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+	if err := scanner.Err(); err != nil {
+		t.Logf("WARNING: error reading node log file: %v", err)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/logscanner_test.go]: fixed

File: test/integration/reject_exception_test.go [hunk 8]
```diff
@@ -32,12 +34,27 @@ func (s *RejectExceptionSuite) TearDownSuite() {
...
 
+func (s *RejectExceptionSuite) SetupTest() {
+	s.StartLogCapture()
+	s.appNames = nil
+}
+
+func (s *RejectExceptionSuite) TearDownTest() {
+	for _, name := range s.appNames {
+		s.T().Logf("Disabling application %s", name)
+		_ = disableApplication(s.ctx, name)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/reject_exception_test.go]: fixed

File: test/integration/echo_prt_test.go [hunk 9]
```diff
@@ -41,15 +43,29 @@ func (s *EchoPrtSuite) TearDownSuite() {
...
 
+func (s *EchoPrtSuite) SetupTest() {
+	s.StartLogCapture()
+	s.appName = ""
+}
+
+func (s *EchoPrtSuite) TearDownTest() {
+	if s.appName != "" {
+		s.T().Logf("Disabling application %s", s.appName)
+		_ = disableApplication(s.ctx, s.appName)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/echo_prt_test.go]: fixed

File: test/integration/multi_app_test.go [hunk 10]
```diff
@@ -65,6 +66,20 @@ func (s *MultiAppSuite) TearDownSuite() {
...
 
+func (s *MultiAppSuite) SetupTest() {
+	s.StartLogCapture()
+}
+
+func (s *MultiAppSuite) TearDownTest() {
+	for _, name := range []string{s.app1Name, s.app2Name} {
+		if name != "" {
+			s.T().Logf("Disabling application %s", name)
+			_ = disableApplication(s.ctx, name)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/multi_app_test.go]: fixed

File: test/integration/reject_exception_prt_test.go [hunk 11]
```diff
@@ -42,14 +44,29 @@ func (s *RejectExceptionPrtSuite) TearDownSuite() {
...
 
+func (s *RejectExceptionPrtSuite) SetupTest() {
+	s.StartLogCapture()
+	s.appNames = nil
+}
+
+func (s *RejectExceptionPrtSuite) TearDownTest() {
+	for _, name := range s.appNames {
+		s.T().Logf("Disabling application %s", name)
+		_ = disableApplication(s.ctx, name)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/reject_exception_prt_test.go]: fixed

File: internal/inspect/inspect_test.go [hunk 12]
```diff
@@ -143,6 +143,58 @@ func (s *InspectSuite) TestPostNoApp() {
...
+	case <-time.After(TestTimeout):
+		s.FailNow("timed out waiting for HttpService to be ready")
+	}
+
+	// Query by name
+	resp, err := http.Post(fmt.Sprintf("http://%s/inspect/%s", s.ServiceAddr, app.Name),
+		"application/octet-stream",
+		bytes.NewBuffer([]byte("hello")))
+	s.Require().Nil(err)
+	s.Equal(http.StatusServiceUnavailable, resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply internal/inspect/inspect_test.go]: fixed

File: internal/inspect/inspect_test.go [hunk 13]
```diff
@@ -143,6 +143,58 @@ func (s *InspectSuite) TestPostNoApp() {
...
+
+	// Query by address
+	resp, err = http.Post(fmt.Sprintf("http://%s/inspect/%s", s.ServiceAddr, app.IApplicationAddress.Hex()),
+		"application/octet-stream",
+		bytes.NewBuffer([]byte("hello")))
+	s.Require().Nil(err)
+	s.Equal(http.StatusServiceUnavailable, resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply internal/inspect/inspect_test.go]: fixed

File: test/integration/echo_authority_test.go [hunk 14]
```diff
@@ -31,16 +34,85 @@ func (s *EchoAuthoritySuite) TearDownSuite() {
...
 
+func (s *EchoAuthoritySuite) SetupTest() {
+	s.StartLogCapture()
+	s.appName = ""
+}
+
+func (s *EchoAuthoritySuite) TearDownTest() {
+	if s.appName != "" {
+		s.T().Logf("Disabling application %s", s.appName)
+		_ = disableApplication(s.ctx, s.appName)
```
[the remark this answers is not in the document]
Comment (vfusco) [reply test/integration/echo_authority_test.go]: fixed

**Review by mpolitzer — APPROVED** [review]
